pkg/adt/interval_tree.go

```diff
@@ -171,45 +171,60 @@ type IntervalValue struct {
 type IntervalTree struct {
 	root  *intervalNode
 	count int
+
+	// red-black NIL node, can not use golang nil instead
```

Comment:
can you explain why we cannot use golang nil to represent the NIL node? If we do need a NIL node instance for comparison, shall we make it a package global variable so that we do not need to pass it around?

Reply:
@xiang90 See the issue #10877 , when delete node 11， if use golang nil instead of rbtree NIL node,  the tree is not rbtree.

Because of delete node 11, the NIL node should be coloring `double black`, but use golang nil will nothing to do.

Reply:
@xiang90  
Because of the tree is also Binary Search Tree, so the code all cases are right.
![image](https://user-images.githubusercontent.com/6369390/62094222-ec4c7c80-b2ae-11e9-9387-71a1c357387d.png)
The Introduction to algorithms tell us when delete leaf `black node`,  then NIL node will be `double black`. 

Red-black tree nature 5: For each node, the same number of black nodes are included on the path of the node nodes after reaching the node.

Reply:
@xiang90 The NIL node can not a package global variable.  It's a pointer variable, there are many trees.